File: src/SemanticTypeForgers/forger.php
<?php

namespace SemanticTypeForgers;

class AtomicBehavior {
    private $name;

    public function __construct($name) {
        $this->name = $name;
    }

    public function setSemanticType($v) {
        return $v;
    }

    public function getPrimitiveType($v) {
        return $v;
    }

    public function validate($value) {
        if ($this->name === 'PersonEmail') {
            $prim = $this->convertToPrimitive($value);
            if (is_string($prim)) {
                return (bool) preg_match('/^[a-zA-Z0-9_.+\-]+@[a-zA-Z0-9\-]+\.[a-zA-Z0-9\-.]+$/', $prim);
            }
            return false;
        }
        return false;
    }

    private function normalizeString($val) {
        return strtolower(trim($val));
    }

    private function primitiveStringToValue($val) {
        $trimmed = trim($val);
        $lower = $this->normalizeString($val);

        if ($trimmed === "") return $val;
        if ($val === "{}" || $val === "[]") return 0;
        if ($lower === "true") return true;
        if ($lower === "false") return false;
        if ($lower === "null" || $lower === "undefined") return null;

        if (preg_match('/^[+-]?(?:(?:\d+\.?\d*)|(?:\.\d+))(?:e[+-]?\d+)?$/i', $trimmed)) {
            if (strpos($trimmed, '.') !== false || strpos($lower, 'e') !== false) {
                return (float)$trimmed;
            } else {
                return (int)$trimmed;
            }
        }

        return $val;
    }

    private function unwrap($val, &$seen, $depth) {
        if ($depth > 100) {
            throw new \Exception("convertToPrimitive: recursive unwrap limit exceeded");
        }

        if ($val === null) return null;

        if (is_bool($val) || is_int($val) || is_float($val)) return $val;

        if (is_string($val)) return $this->primitiveStringToValue($val);

        if (is_array($val)) {
            if (empty($val)) {
                if (array_keys($val) !== range(0, count($val) - 1)) {
                    return 0; // Assoc array / Object equivalent
                }
                return null;
            }
            return $this->unwrap(reset($val), $seen, $depth + 1);
        }

        if (is_object($val)) {
            $id = spl_object_id($val);
            if (isset($seen[$id])) {
                throw new \Exception("convertToPrimitive: circular reference detected");
            }
            $seen[$id] = true;

            if (method_exists($val, 'getPrimitiveType')) {
                return $this->unwrap($val->getPrimitiveType(), $seen, $depth + 1);
            }

            if ($val instanceof \DateTimeInterface) {
                return $val->getTimestamp() * 1000; // rough JS timestamp equivalent
            }

            $props = get_object_vars($val);
            if (empty($props)) return 0;

            return $this->unwrap(reset($props), $seen, $depth + 1);
        }

        return null;
    }

    public function convertToPrimitive($value) {
        $seen = [];
        return $this->unwrap($value, $seen, 0);
    }

    public function forge($v) {
        if ($this->name === 'PersonEmail' && !$this->validate($v)) {
            throw new \Exception("Validation failed for SemanticType: {$this->name}");
        }
        return $v;
    }

    public function proccessValue($value) {
        if ($this->validate($value)) return $value;

        if (is_array($value) && array_keys($value) !== range(0, count($value) - 1)) {
            $productPrice = $this->convertToPrimitive($value['productPrice'] ?? null);
            $productDiscount = $this->convertToPrimitive($value['productDiscount'] ?? null) ?: 0;
            $deliveryPrice = $this->convertToPrimitive($value['deliveryPrice'] ?? null) ?: 0;
            $paymentFees = $this->convertToPrimitive($value['paymentFees'] ?? null) ?: 0;

            if ($productPrice !== null && (is_int($productPrice) || is_float($productPrice))) {
                $finalPrice = $productPrice - $productDiscount + $deliveryPrice + $paymentFees;
                return $this->forge($finalPrice);
            }
        } elseif (is_object($value)) {
            $productPrice = $this->convertToPrimitive($value->productPrice ?? null);
            $productDiscount = $this->convertToPrimitive($value->productDiscount ?? null) ?: 0;
            $deliveryPrice = $this->convertToPrimitive($value->deliveryPrice ?? null) ?: 0;
            $paymentFees = $this->convertToPrimitive($value->paymentFees ?? null) ?: 0;

            if ($productPrice !== null && (is_int($productPrice) || is_float($productPrice))) {
                $finalPrice = $productPrice - $productDiscount + $deliveryPrice + $paymentFees;
                return $this->forge($finalPrice);
            }
        }

        $primitiveVal = $this->convertToPrimitive($value);
        return $this->forge($primitiveVal);
    }
}

File: src/SemanticTypeForgers/forger.py
import re
from datetime import datetime

def AtomicBehavior(name):
    def getPrimitiveType(v):
        return v

    def validate(value):
        return False

    def convertToPrimitive(value):
        def normalizeString(val):
            return val.strip().lower()

        def primitiveStringToValue(val):
            trimmed = val.strip()
            lower = normalizeString(val)

            if trimmed == "":
                return val
            if val == "{}" or val == "[]":
                return 0
            if lower == "true":
                return True
            if lower == "false":
                return False
            if lower == "null":
                return None
            if lower == "undefined":
                return None

            isNumberLike = re.match(r'^[+-]?(?:(?:\d+\.?\d*)|(?:\.\d+))(?:e[+-]?\d+)?$', trimmed, re.IGNORECASE)
            if isNumberLike:
                try:
                    if '.' in trimmed or 'e' in trimmed.lower():
                        return float(trimmed)
                    return int(trimmed)
                except ValueError:
                    pass
            return val

        def unwrap(val, seen=None, depth=0):
            if seen is None:
                seen = set()

            if depth > 100:
                raise Exception("convertToPrimitive: recursive unwrap limit exceeded")

            if val is None:
                return val

            if not val and not isinstance(val, (dict, list, set, tuple, object)) and type(val) not in (int, float, bool, str):
                return val

            if type(val) in (int, float, bool):
                return val

            if isinstance(val, str):
                return primitiveStringToValue(val)

            if callable(val):
                return unwrap(val(), seen, depth + 1)

            if isinstance(val, (list, tuple)):
                if len(val) == 0:
                    return None
                return unwrap(val[0], seen, depth + 1)

            if isinstance(val, dict):
                if not val:
                    return 0
                values = list(val.values())
                if len(values) == 0:
                    return None
                return unwrap(values[0], seen, depth + 1)

            if hasattr(val, "__class__") and val.__class__.__name__ not in ("int", "float", "str", "bool", "list", "dict", "tuple", "set", "NoneType"):
                if id(val) in seen:
                    raise Exception("convertToPrimitive: circular reference detected")
                seen.add(id(val))

                if hasattr(val, "getPrimitiveType") and callable(getattr(val, "getPrimitiveType")):
                    return unwrap(val.getPrimitiveType(), seen, depth + 1)

                if isinstance(val, datetime):
                    return unwrap(val.timestamp() * 1000, seen, depth + 1)

                if hasattr(val, "__dict__"):
                    props = [v for k, v in val.__dict__.items() if not k.startswith('_')]
                    if not props:
                        return 0
                    return unwrap(props[0], seen, depth + 1)

            return None

        return unwrap(value)

    def forge(v):
        return v

    def proccessValue(value):
        if validate(value):
            return value

        if value is not None and isinstance(value, dict):
            obj = value
            productPrice = convertToPrimitive(obj.get("productPrice"))
            productDiscount = convertToPrimitive(obj.get("productDiscount")) or 0
            deliveryPrice = convertToPrimitive(obj.get("deliveryPrice")) or 0
            paymentFees = convertToPrimitive(obj.get("paymentFees")) or 0

            if productPrice is not None and isinstance(productPrice, (int, float)):
                finalPrice = productPrice - productDiscount + deliveryPrice + paymentFees
                return forge(finalPrice)
        elif value is not None and hasattr(value, "__dict__"):
            productPrice = convertToPrimitive(getattr(value, "productPrice", None))
            productDiscount = convertToPrimitive(getattr(value, "productDiscount", None)) or 0
            deliveryPrice = convertToPrimitive(getattr(value, "deliveryPrice", None)) or 0
            paymentFees = convertToPrimitive(getattr(value, "paymentFees", None)) or 0

            if productPrice is not None and isinstance(productPrice, (int, float)):
                finalPrice = productPrice - productDiscount + deliveryPrice + paymentFees
                return forge(finalPrice)

        primitiveVal = convertToPrimitive(value)
        return forge(primitiveVal)

    return {
        "setSemanticType": lambda v: v,
        "getPrimitiveType": getPrimitiveType,
        "validate": validate,
        "convertToPrimitive": convertToPrimitive,
        "forge": forge,
        "proccessValue": proccessValue,
    }
